dashboard/app/models/user.rb

```diff
@@ -769,13 +769,39 @@ def track_script_progress(script)
     end
   end
 
-  # returns whether a new level has been completed
-  def track_level_progress(script_level, new_result)
+  # returns whether a new level has been completed and asynchronously enqueues an operation
+  # to update the level progress.
+  def track_level_progress_async(script_level, new_result)
+    new_level_completed = false
+    level_id = script_level.level_id
+    script_id = script_level.script_id
+    user_level = UserLevel.where(user_id: self.id,
```

Comment:
It looks like you don't use this user_level (to optionally create a new one) here anymore so I think you could simplify by making this `first` instead of `first_or_initialize` and then checking `user_level.nil? || !user_level.passing?` or even `Activity.passing?(user_level.try(:best_result))` (which has a minor advantage of using the same method for the same meaning) in L794.

Could even call this `old_user_level` or something now that it's not also used to make the new user_level so then you don't need the "user_level is the old result" comment